skfem/assembly/basis/facet_basis.py

```diff
@@ -108,10 +115,110 @@ def default_parameters(self):
                 'h': self.mesh_parameters(),
                 'n': self.normals}
 
-    def global_coordinates(self) -> ndarray:
+    def global_coordinates(self) -> DiscreteField:
         return DiscreteField(self.mapping.G(self.X, find=self.find))
 
-    def mesh_parameters(self) -> ndarray:
+    def mesh_parameters(self) -> DiscreteField:
         return DiscreteField((np.abs(self.mapping.detDG(self.X, self.find))
                               ** (1. / (self.mesh.dim() - 1.)))
                              if self.mesh.dim() != 1 else np.array([0.]))
+
+    def _trace(self,
+               x: ndarray,
+               elem: Element) -> Tuple[ndarray, ndarray, ndarray]:
+        """Trace mesh reindexing and basis projection."""
+        from skfem.utils import project
+
+        fbasis = FacetBasis(self.mesh,
+                            elem,
+                            facets=self.find,
+                            quadrature=(self.X, self.W))
+        I = fbasis.get_dofs(self.find).all()
+        if len(I) == 0:  # special case: no facet DOFs
+            if fbasis.dofs.interior_dofs.shape[0] > 1:
+                # no one-to-one restriction: requires interpolation
+                raise NotImplementedError
+            # special case: piecewise constant elem
+            I = fbasis.dofs.interior_dofs[:, self.tind].flatten()
+        y = project(x, self, fbasis, I=I)
+
+        # build connectivity for lower dimensional mesh
+        ix = self.mesh.facets[:, self.find]
+        ixuniq = np.unique(ix)
+        b = np.zeros(np.max(ix) + 1, dtype=np.int64)
+        b[ixuniq] = np.arange(len(ixuniq), dtype=np.int64)
+
+        return self.mesh.p[:, ixuniq], b[ix], y
+
```

Comment:
`_trace` looks as though it could be split into two functions, one returning the points and cells and the other returning the values. The parts of the body either side of the &lsquo;build connectivity&rsquo; comment are independent.

Reply:
Yep.

```python
 def _trace_points_cells(self) -> Tuple[ndarray, ndarray, ndarray]:
        """Trace mesh reindexing and basis projection."""

        # build connectivity for lower dimensional mesh
        ix = self.mesh.facets[:, self.find]
        ixuniq = np.unique(ix)
        b = np.zeros(np.max(ix) + 1, dtype=np.int64)
        b[ixuniq] = np.arange(len(ixuniq), dtype=np.int64)

        return self.mesh.p[:, ixuniq], b[ix]

    def _trace_values(
        self, x: ndarray, elem: Element
    ) -> Tuple[ndarray, ndarray, ndarray]:
        """Trace mesh reindexing and basis projection."""
        from skfem.utils import project

        fbasis = FacetBasis(
            self.mesh, elem, facets=self.find, quadrature=(self.X, self.W)
        )
        I = fbasis.get_dofs(self.find).all()
        if len(I) == 0:  # special case: no facet DOFs
            if fbasis.dofs.interior_dofs.shape[0] > 1:
                # no one-to-one restriction: requires interpolation
                raise NotImplementedError
            # special case: piecewise constant elem
            I = fbasis.dofs.interior_dofs[:, self.tind].flatten()
        y = project(x, self, fbasis, I=I)

        return y

    def _trace(self, x: ndarray, elem: Element) -> Tuple[ndarray, ndarray, ndarray]:
        """Trace mesh reindexing and basis projection."""
        from skfem.utils import project

        return *self._trace_points_cells(), self._trace_values(x, elem)
```

Not to say that that's how it should be written, just to demonstrate the factorization.

Reply:
Apart from the use of `self.find`, `_trace_points_cells` doesn't have too much to do with the trace.  It sort of looks like some kind of submesh function, if it were passed `ix`.  I wonder whether such a function (or rather method of class Mesh) would be useful for anything else.  Untested:

```python
def submesh(self: Mesh, ix: ndarray) -> Tuple[ndarray, ndarray]:
    ixuniq = np.unique(ix)
    b = np.zeros(np.max(ix) + 1, dtype=np.int64)
    b[ixuniq] = np.arange(len(ixuniq), dtype=np.int64)
   
    return self.p[:, ixuniq], b[ix]
```

Reply:
It seems that similar logic to `submesh` is used at least in `Mesh.remove_elements`. Thus, it could make sense to have it in `Mesh`. The same function could work for subsets of elements, facets or edges. I'll look into refactoring it.